tactics/tactics.ml

```diff
@@ -1645,6 +1645,9 @@ let make_projection env sigma params cstr sign elim i n c u =
       | None -> None
   in elim
 
+(* Used for detecting failure of the intro pattern with "apply in" *)
+exception FailureAfterApply of exn
```

Comment:
Wrapping exceptions is deemed dangerous in the Coq source code. I wonder whether this shouldn't be done in a more functional way, e.g. returning an algebraic type in the related functions.